pkg/scheduler/framework/v1alpha1/waiting_pods_map.go

```diff
@@ -69,26 +71,59 @@ func (m *waitingPodsMap) iterate(callback func(WaitingPod)) {
 
 // waitingPod represents a pod waiting in the permit phase.
 type waitingPod struct {
-	pod *v1.Pod
-	s   chan *Status
+	pod            *v1.Pod
+	pendingPlugins map[string]*time.Timer
+	s              chan *Status
 }
 
 // newWaitingPod returns a new waitingPod instance.
-func newWaitingPod(pod *v1.Pod) *waitingPod {
-	return &waitingPod{
+func newWaitingPod(pod *v1.Pod, pluginsMaxWaitTime map[string]time.Duration) *waitingPod {
+	wp := &waitingPod{
 		pod: pod,
 		s:   make(chan *Status),
 	}
+
+	pendingPlugins := make(map[string]*time.Timer)
+	for k, v := range pluginsMaxWaitTime {
+		plugin, waitTime := k, v
+		pendingPlugins[plugin] = time.AfterFunc(waitTime, func() {
+			msg := fmt.Sprintf("rejected due to timeout after waiting %v at plugin %v",
+				waitTime, plugin)
+			wp.Reject(msg)
+		})
+	}
+	wp.pendingPlugins = pendingPlugins
```

Comment:
Why are you using an auxiliar map and not directly the waitingPod field?

Reply:
Updated to use waitingPod filed directly.